server/src/main/java/org/opensearch/cluster/metadata/ComposableIndexTemplate.java

```diff
@@ -264,51 +266,74 @@ public String toString() {
 
     public static class DataStreamTemplate implements Writeable, ToXContentObject {
 
-        private static final ObjectParser<DataStreamTemplate, Void> PARSER = new ObjectParser<>(
+        private static final ParseField TIMESTAMP_FIELD_FIELD = new ParseField("timestamp_field");
+
+        private static final ConstructingObjectParser<DataStreamTemplate, Void> PARSER = new ConstructingObjectParser<>(
             "data_stream_template",
-            DataStreamTemplate::new
+            true,
+            args -> new DataStreamTemplate((TimestampField) args[0])
         );
 
+        static {
+            PARSER.declareObject(ConstructingObjectParser.optionalConstructorArg(), TimestampField.PARSER, TIMESTAMP_FIELD_FIELD);
+        }
+
+        private final TimestampField timestampField;
+
         public DataStreamTemplate() {
+            this(DataStreamFieldMapper.Defaults.TIMESTAMP_FIELD);
+        }
+
+        public DataStreamTemplate(TimestampField timestampField) {
+            this.timestampField = timestampField;
         }
 
-        public String getTimestampField() {
-            return FIXED_TIMESTAMP_FIELD;
+        public DataStreamTemplate(StreamInput in) throws IOException {
+            this.timestampField = in.readOptionalWriteable(TimestampField::new);
         }
 
-        DataStreamTemplate(StreamInput in) {
-            this();
+        public TimestampField getTimestampField() {
+            return timestampField == null ? DataStreamFieldMapper.Defaults.TIMESTAMP_FIELD : timestampField;
         }
 
         /**
          * @return a mapping snippet for a backing index with `_data_stream_timestamp` meta field mapper properly configured.
          */
         public Map<String, Object> getDataStreamMappingSnippet() {
-            // _data_stream_timestamp meta fields default to @timestamp:
-            return singletonMap(MapperService.SINGLE_MAPPING_NAME, singletonMap("_data_stream_timestamp",
-                singletonMap("enabled", true)));
+            return singletonMap(
+                MapperService.SINGLE_MAPPING_NAME, singletonMap(
+                    "_data_stream_timestamp", unmodifiableMap(of(
+                        "enabled", true,
+                        "timestamp_field", getTimestampField().toMap()
+                    ))
+                )
+            );
         }
 
         @Override
         public void writeTo(StreamOutput out) throws IOException {
+            out.writeOptionalWriteable(timestampField);
```

Comment:
Also be sure to wrap the companion writeTo in a version check.
```suggestion
            if (out.getVersion().onOrAfter(Version.V1_0_0)) {
                out.writeOptionalWriteable(timestampField);
            }    
```

Reply:
Thanks for pointing this out. I have added this check along with UTs.